### PR TITLE
Allow multiple tray_output directives.

### PR DESCRIPTION
**Pull request by Airblader**
This patch introduces the possibility to specify the tray_output directive
multiple times. All values will be used by i3bar, in the order they are
given.

This way, a single bar configuration can be used for several machines with
internal output names "eDP1" and "LVDS-0" by specifying tray_output for both.
Any external output (e.g., "DP-0") will still not receive the tray. The same
effect can be achieved by using "primary", but forces the user to couple the
tray display to the primary output which may not be desirable behavior.

relates to #555

### CODE REVIEW

**Comment by Airblader** [comment]
@stapelberg You can review already, but I'd like to test this PR a bit more over the weekend with some configurations.

File: i3bar/src/xcb.c [hunk 1]
```diff
@@ -678,25 +678,45 @@ static void handle_client_message(xcb_client_message_event_t *event) {
...
-            SLIST_FOREACH(walk, outputs, slist) {
-                if (!walk->active)
-                    continue;
-                if (config.tray_output) {
-                    if ((strcasecmp(walk->name, config.tray_output) != 0) &&
-                        (!walk->primary || strcasecmp("primary", config.tray_output) != 0))
+            i3_output *output = NULL;
+            i3_output *walk = NULL;
+            tray_output_t *tray_output = NULL;
+            TAILQ_FOREACH(tray_output, &(config.tray_outputs), tray_outputs) {
```
Comment (Airblader) [inline i3bar/src/xcb.c]: We need to iterate over `tray_outputs` in the outer loop to be independent of the order of `outputs`. For example, say we have outputs `eDP1` and `HDMI1` and configure both as `tray_output` in that order. If we iterate over outputs in the outer loop, we would always choose whichever output is first in that list as an assignment for it exists. The order of `tray_output` assignments needs to be prioritized.

File: i3bar/src/xcb.c [hunk 2]
```diff
@@ -1676,20 +1696,36 @@ void reconfig_windows(bool redraw_bars) {
...
+
+                /* If no tray_output was specified, we go ahead and initialize the tray as
+                 * we will be using the first available output. */
+                if (TAILQ_EMPTY(&(config.tray_outputs)))
+                    init_tray();
+
+                /* If a or many tray_output were specified, we ensure that at least one of
+                 * them is actually an output managed by this instance. */
+                tray_output_t *tray_output;
+                TAILQ_FOREACH(tray_output, &(config.tray_outputs), tray_outputs) {
```
Comment (Airblader) [inline i3bar/src/xcb.c]: Same argument as above. It's actually pretty ugly that we restate the entire logic here, we should probably clean this up. In fact, there's no need to decide where to put a tray icon for every docking request, is there? We should probably decide where the tray is on startup (and react to changes), but then cache this information for docking requests.

File: src/config.c [hunk 3]
```diff
@@ -111,7 +111,6 @@ void load_configuration(xcb_connection_t *conn, const char *override_configpath,
             for (int c = 0; c < barconfig->num_outputs; c++)
                 free(barconfig->outputs[c]);
             FREE(barconfig->outputs);
-            FREE(barconfig->tray_output);
```
Comment (Airblader) [inline src/config.c]: @stapelberg I'm a bit lost here. I figure we need to clear the tailq of tray_outputs here now, but doing so in the following manner causes a config reload to freeze i3 in the "FREE" call.

````
while (!TAILQ_EMPTY(&(barconfig->tray_outputs))) {
    struct tray_output_t *tray_output = TAILQ_FIRST(&(barconfig->tray_outputs));
    FREE(tray_output->output); // freezes here
    TAILQ_REMOVE(…);
    FREE(tray_output);
}
```

I don't understand why this happens (and why it freezes instead of crashing).

**Comment by Airblader** [comment]
Sorry, I should've mentioned that I tested it already now. I tried as many scenarios as I could think of:

* No directive.
* Only `tray_output none`
* Only `tray_output primary`
* `tray_output foobar`, `tray_output eDP1`
* `tray_output eDP1`, `tray_output foobar`
* `tray_output HDM1`, `tray_output eDP1`
* `tray_output eDP1`, `tray_output HDM1`
* `tray_output foobar`, `tray_output eDP1`, `tray_output HDM1`
* `tray_output foobar`, `tray_output HDM1`, `tray_output eDP1`
* Some more I can't think of right now.

File: i3bar/src/config.c [hunk 4]
```diff
@@ -198,10 +213,13 @@ static int config_string_cb(void *params_, const unsigned char *val, size_t _len
         return 1;
     }
 
+    /* We keep the old single tray_output working for users who only restart i3bar
+     * after updating. */
     if (!strcmp(cur_key, "tray_output")) {
-        DLOG("tray_output %.*s\n", len, val);
-        FREE(config.tray_output);
-        sasprintf(&config.tray_output, "%.*s", len, val);
+        ELOG("Found deprecated key tray_output %.*s.\n", len, val);
```
Comment (stapelberg) [inline i3bar/src/config.c]: nit: I think this should be DLOG instead of ELOG. It’s not an error, it’s the expected migration path.

File: i3bar/src/xcb.c (hunk 1 above)
Comment (stapelberg) [inline i3bar/src/xcb.c]: Agreed. Can you put that into a comment in the file?

File: i3bar/src/xcb.c [hunk 5]
```diff
@@ -1676,20 +1696,36 @@ void reconfig_windows(bool redraw_bars) {
...
-                SLIST_FOREACH(output, outputs, slist) {
-                    if (strcasecmp(output->name, tray_output) == 0 ||
-                        (strcasecmp(tray_output, "primary") == 0 && output->primary)) {
-                        init_tray();
-                        break;
+            /* Unless "tray_output none" was specified, we need to initialize the tray. */
+            const char *first = (TAILQ_EMPTY(&(config.tray_outputs))) ? SLIST_FIRST(outputs)->name : TAILQ_FIRST(&(config.tray_outputs))->output;
+            if (!tray_configured && strcasecmp(first, "none") != 0) {
+                /* We do a sanity check here to ensure that this i3bar instance actually handles
+                 * the output on which the tray should appear. */
```
Comment (stapelberg) [inline i3bar/src/xcb.c]: This modification of the comment removes the example, but I like examples. Please add it back :).

File: i3bar/src/xcb.c [hunk 6]
```diff
@@ -1676,20 +1696,36 @@ void reconfig_windows(bool redraw_bars) {
...
+
+                /* If no tray_output was specified, we go ahead and initialize the tray as
+                 * we will be using the first available output. */
+                if (TAILQ_EMPTY(&(config.tray_outputs)))
+                    init_tray();
+
+                /* If a or many tray_output were specified, we ensure that at least one of
```
Comment (stapelberg) [inline i3bar/src/xcb.c]: nit: I think “If one or more tray_outputs were specified” is more idiomatic. But double-check with your girlfriend, if she’s around.

Comment (Airblader) [inline i3bar/src/xcb.c]: No, you're right. I must've been tired… :)

File: src/config.c (hunk 3 above)
Comment (stapelberg) [inline src/config.c]: Strange. The code looks correct to me at first glance. Try attaching gdb and single-stepping through the code to see what’s going on.

File: i3bar/src/xcb.c (hunk 2 above)
Comment (stapelberg) [inline i3bar/src/xcb.c]: Okay. Do you want to do that refactoring in a separate PR, as it’s independent of allowing multiple tray_output directives?

Comment (Airblader) [inline i3bar/src/xcb.c]: Yes, we should do this separately. I'll open a ticket.

File: src/config.c (hunk 3 above)
Comment (Airblader) [inline src/config.c]: Yeah, that's what I did and how I learned that it just hangs in the `FREE` call. I can't make sense of it, to be honest, and was hoping you'd have an idea. :) I'll try looking into it again (fresh day, fresh mind) and at least save the backtrace.

Comment (stapelberg) [inline src/config.c]: Did you check whether valgrind finds any memory corruption with your PR?

Comment (Airblader) [inline src/config.c]: The problem with that is that running i3 under valgrind gives me hundreds of results anyway, so it's pretty difficult to say whether there's an entry more or not (am I doing something wrong? My knowledge about valgrind is pretty much just "run it and look at what it says").

What I noticed, though, is that we don't free things like `autostarts` and `autostarts_always` here either. Do you know if there's a reason for that?

Comment (Airblader) [inline src/config.c]: I did it anyway and at least I couldn't find anything among the dozens of "definitely lost" warnings related to this. But I also can't explain why we wouldn't lose memory here. As mentioned above, the same goes for the autostart command queues, though.

Comment (stapelberg) [inline src/config.c]: > The problem with that is that running i3 under valgrind gives me hundreds of results anyway, so it's pretty difficult to say whether there's an entry more or not (am I doing something wrong? My knowledge about valgrind is pretty much just "run it and look at what it says").

Are you using the --valgrind flag of complete-run.pl? If not, have a look at how testcases/lib/SocketActivation.pm
 calls valgrind. That should give you a more manageable report.

> What I noticed, though, is that we don't free things like autostarts and autostarts_always here either. Do you know if there's a reason for that?

That’s an oversight. I think currently, they will just grow indefinitely. I think nobody noticed because the autostart directive is only relevant when i3 starts.

Comment (Airblader) [inline src/config.c]: > Are you using the --valgrind flag of complete-run.pl?

No, I was literally running `valgrind --leak-check=full i3` from my `.xinitrc` on my machine. :) I'll look into the test.

> That’s an oversight. I think currently, they will just grow indefinitely. 

But then I'd expect `exec_always` commands to be executed multiple times, one more time each time i3 is restarted. Using `exec_always echo x >> /tmp/foo` I cannot confirm this, however – it runs exactly once on each restart.

Comment (Airblader) [inline src/config.c]: @stapelberg I ran i3 under valgrind now and reloaded the config a bunch of times. Indeed, I can see this:

````
===15547== 6 bytes in 1 blocks are definitely lost in loss record 60 of 1,098                                                                              
==15547==    at 0x4C2A987: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15547==    by 0x44AC38: scalloc (safewrappers.c:31)
==15547==    by 0x428442: handle_get_bar_config (ipc.c:924)
==15547==    by 0x428A66: ipc_receive_message (ipc.c:1099)
==15547==    by 0x76DC5D2: ev_invoke_pending (in /usr/lib/libev.so.4.0.0)
==15547==    by 0x76DFC0D: ev_run (in /usr/lib/libev.so.4.0.0)
==15547==    by 0x417BDA: ev_loop (ev.h:828)
==15547==    by 0x41B60D: main (main.c:852)
==15547==
==15547== 8 bytes in 1 blocks are definitely lost in loss record 107 of 1,098
==15547==    at 0x4C28C10: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15547==    by 0x7FFEE09: strdup (in /usr/lib/libc-2.22.so)
==15547==    by 0x44ACD9: sstrdup (safewrappers.c:45)
==15547==    by 0x41F9F6: cfg_bar_color_single (config_directives.c:594)
==15547==    by 0x43E9EE: GENERATED_call (GENERATED_config_call.h:198)
==15547==    by 0x43FB34: next_state (config_parser.c:208)
==15547==    by 0x4401AD: parse_config (config_parser.c:390)
==15547==    by 0x441A28: parse_file (config_parser.c:977)
==15547==    by 0x40ECC2: parse_configuration (config.c:65)
==15547==    by 0x40F81B: load_configuration (config.c:211)
==15547==    by 0x41A6A4: main (main.c:515)
==15547==
==15547== 8 bytes in 1 blocks are definitely lost in loss record 108 of 1,098
==15547==    at 0x4C28C10: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15547==    by 0x7FFEE09: strdup (in /usr/lib/libc-2.22.so)
==15547==    by 0x44ACD9: sstrdup (safewrappers.c:45)
==15547==    by 0x41F9F6: cfg_bar_color_single (config_directives.c:594)
==15547==    by 0x43E9EE: GENERATED_call (GENERATED_config_call.h:198)
==15547==    by 0x43FB34: next_state (config_parser.c:208)
==15547==    by 0x4401AD: parse_config (config_parser.c:390)
==15547==    by 0x441A28: parse_file (config_parser.c:977)
==15547==    by 0x40ECC2: parse_configuration (config.c:65)
==15547==    by 0x40F81B: load_configuration (config.c:211)
==15547==    by 0x4324E0: cmd_reload (commands.c:1595)
==15547==    by 0x435E5E: GENERATED_call (GENERATED_command_call.h:510)
==15547==    by 0x4365D9: next_state (commands_parser.c:187)
==15547==    by 0x436AA9: parse_command (commands_parser.c:308)
==15547==    by 0x425660: handle_command (ipc.c:92)
==15547==    by 0x428A66: ipc_receive_message (ipc.c:1099)
==15547==    by 0x76DC5D2: ev_invoke_pending (in /usr/lib/libev.so.4.0.0)
==15547==    by 0x76DFC0D: ev_run (in /usr/lib/libev.so.4.0.0)
==15547==    by 0x417BDA: ev_loop (ev.h:828)
==15547==    by 0x41B60D: main (main.c:852)
==15547==
==15547== 29 (24 direct, 5 indirect) bytes in 1 blocks are definitely lost in loss record 386 of 1,098
==15547==    at 0x4C2A987: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15547==    by 0x44AC38: scalloc (safewrappers.c:31)
==15547==    by 0x41F921: cfg_bar_tray_output (config_directives.c:581)
==15547==    by 0x43EF6C: GENERATED_call (GENERATED_config_call.h:406)
==15547==    by 0x43FB34: next_state (config_parser.c:208)
==15547==    by 0x4401AD: parse_config (config_parser.c:390)
==15547==    by 0x441A28: parse_file (config_parser.c:977)
==15547==    by 0x40ECC2: parse_configuration (config.c:65)
==15547==    by 0x40F81B: load_configuration (config.c:211)
==15547==    by 0x41A6A4: main (main.c:515)
==15547==
==15547== 29 (24 direct, 5 indirect) bytes in 1 blocks are definitely lost in loss record 387 of 1,098
==15547==    at 0x4C2A987: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15547==    by 0x44AC38: scalloc (safewrappers.c:31)
==15547==    by 0x41F921: cfg_bar_tray_output (config_directives.c:581)
==15547==    by 0x43EF6C: GENERATED_call (GENERATED_config_call.h:406)
==15547==    by 0x43FB34: next_state (config_parser.c:208)
==15547==    by 0x4401AD: parse_config (config_parser.c:390)
==15547==    by 0x441A28: parse_file (config_parser.c:977)
==15547==    by 0x40ECC2: parse_configuration (config.c:65)
==15547==    by 0x40F81B: load_configuration (config.c:211)
==15547==    by 0x4324E0: cmd_reload (commands.c:1595)
==15547==    by 0x435E5E: GENERATED_call (GENERATED_command_call.h:510)
==15547==    by 0x4365D9: next_state (commands_parser.c:187)
==15547==    by 0x436AA9: parse_command (commands_parser.c:308)
==15547==    by 0x425660: handle_command (ipc.c:92)
==15547==    by 0x428A66: ipc_receive_message (ipc.c:1099)
==15547==    by 0x76DC5D2: ev_invoke_pending (in /usr/lib/libev.so.4.0.0)
==15547==    by 0x76DFC0D: ev_run (in /usr/lib/libev.so.4.0.0)
==15547==    by 0x417BDA: ev_loop (ev.h:828)
==15547==    by 0x41B60D: main (main.c:852)
==15547==
```

On a slightly different note, I also found this:

````
==15547== 136 bytes in 1 blocks are definitely lost in loss record 843 of 1,098
==15547==    at 0x4C2A987: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15547==    by 0x5357497: xkb_state_new (in /usr/lib/libxkbcommon.so.0.0.0)
==15547==    by 0x40894D: translate_keysyms (bindings.c:317)
==15547==    by 0x41AA11: main (main.c:608)
==15547==
```

Comment (Airblader) [inline src/config.c]: I also just verified that the autostarts lists do in fact grow infinitely on config reloads.

Comment (Airblader) [inline src/config.c]: @stapelberg Could the way we copy the bar configuration after parsing a bar block be the problem?

````
   /* Copy the current (static) structure into a dynamically allocated
     * one, then cleanup our static one. */
    Barconfig *bar_config = scalloc(1, sizeof(Barconfig));
    memcpy(bar_config, &current_bar, sizeof(Barconfig));
    TAILQ_INSERT_TAIL(&barconfigs, bar_config, configs);

    memset(&current_bar, '\0', sizeof(Barconfig));
```

My knowledge on C and memory is a bit too limited to understand whether we'd have a problem here.

Comment (stapelberg) [inline src/config.c]: I’ve pushed 7b502cf44cabd4d01dcadaae3ecb8b43be97730d and 8125b54e09aa2ec06367245d8132241a8ee682ae which were legitimate memleaks.

As for the rest, the line numbers in valgrind don’t match up with the source currently in this pull request or in “next”, so I can’t really understand them.

I think the bar-copying is fine and wouldn’t know why it would interfere with anything.

Comment (Airblader) [inline src/config.c]: Yeah, sorry, that was my fault. I used valgrind on i3-gaps with this PR merged in, that's why the line numbers are incorrect. I'll repeat it with vanilla i3. On a separate not, we'll need to look into pango as well because like 99% of the dozens of loss reports come from it.

Comment (Airblader) [inline src/config.c]: @stapelberg This time with vanilla i3:

````
==8388== 61 (24 direct, 37 indirect) bytes in 1 blocks are definitely lost in loss record 571 of 1,071
==8388==    at 0x4C2A987: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==8388==    by 0x4497D8: scalloc (safewrappers.c:31)
==8388==    by 0x41F4BC: cfg_bar_tray_output (config_directives.c:523)
==8388==    by 0x43E2FB: GENERATED_call (GENERATED_config_call.h:406)
==8388==    by 0x43EBF1: next_state (config_parser.c:208)
==8388==    by 0x43F26A: parse_config (config_parser.c:390)
==8388==    by 0x440AE5: parse_file (config_parser.c:977)
==8388==    by 0x40ECCE: parse_configuration (config.c:65)
==8388==    by 0x40F813: load_configuration (config.c:208)
==8388==    by 0x41A422: main (main.c:515)
==8388== 
==8388== 64 (32 direct, 32 indirect) bytes in 1 blocks are definitely lost in loss record 632 of 1,071
==8388==    at 0x4C2A987: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==8388==    by 0x4497D8: scalloc (safewrappers.c:31)
==8388==    by 0x407F38: mode_from_name (bindings.c:40)
==8388==    by 0x4081FF: configure_binding (bindings.c:95)
==8388==    by 0x41DDE7: cfg_mode_binding (config_directives.c:122)
==8388==    by 0x43E103: GENERATED_call (GENERATED_config_call.h:334)
==8388==    by 0x43EBF1: next_state (config_parser.c:208)
==8388==    by 0x43F26A: parse_config (config_parser.c:390)
==8388==    by 0x440AE5: parse_file (config_parser.c:977)
==8388==    by 0x40ECCE: parse_configuration (config.c:65)
==8388==    by 0x40F813: load_configuration (config.c:208)
==8388==    by 0x41A422: main (main.c:515)
```

Comment (Airblader) [inline src/config.c]: @stapelberg Did you get a chance to look at this? I keep looking at it every day, but I'm really lost on what's happening here.

Comment (Airblader) [inline src/config.c]: I think I got it to work. I'll be frank – I still don't understand what is going on, but I've switched to dynamically allocating memory for the temporary barconfig now and it doesn't crash anymore. I will give it another gdb run, though, to see if the memory leak is really gone.

**Comment by Airblader** [comment]
@stapelberg OK, I've run it through valgrind and can't find any trace of a memory leak in this area anymore. Also seems to work fine for me.

**Comment by Airblader** [comment]
Actually hold on one more time. I did something wrong when testing this. Sorry!

**Comment by Airblader** [comment]
The crash is still there. :( I have written a small minimal example of the problem now which reproduces this and will ask about it on stackoverflow.

**Comment by Airblader** [comment]
http://stackoverflow.com/questions/33397351/nested-tailq-double-free-error

**Comment by stapelberg** [comment]
Sorry, didn’t get a chance so far to reproduce.

Thanks for extracting the minimum example, that makes it so much easier to test. I’ve posted a reply on StackOverflow: http://stackoverflow.com/a/33398441/712014

**Comment by Airblader** [comment]
Thank you, I think this sounds correct. For status blocks we have the same logic, that I implemented some time ago, and I remember fighting with this for a long time back then as well. Ah, the struggles of memory management.

I'll see if I can update the PR later today before my flight to the US tomorrow, otherwise I'll do it from there.

File: src/config_directives.c [hunk 7]
```diff
@@ -516,73 +516,72 @@ CFGFUN(bar_color, const char *colorclass, const char *border, const char *backgr
...
-    /* Copy the current (static) structure into a dynamically allocated
-     * one, then cleanup our static one. */
-    Barconfig *bar_config = scalloc(1, sizeof(Barconfig));
-    memcpy(bar_config, &current_bar, sizeof(Barconfig));
-    TAILQ_INSERT_TAIL(&barconfigs, bar_config, configs);
-
-    memset(&current_bar, '\0', sizeof(Barconfig));
+    TAILQ_INSERT_TAIL(&barconfigs, current_bar, configs);
+    /* Simply reset the pointer, but don't free the resources. */
+    current_bar = NULL;
```
Comment (Airblader) [inline src/config_directives.c]: @stapelberg Simply inserting `current_bar` seems to be easier than copying over all content to a newly allocated instance. I did the same in my tailq example and ran it in valgrind and it didn't produce any memory leaks.

I also valgrinded this PR again now. No more memory leaks, in fact I found two more.

**Comment by Airblader** [comment]
@stapelberg OK, memory leaks are gone and reloading works. Everything should finally be OK.

File: src/config.c [hunk 8]
```diff
@@ -75,20 +75,20 @@ void load_configuration(xcb_connection_t *conn, const char *override_configpath,
         ungrab_all_keys(conn);
 
         struct Mode *mode;
-        Binding *bind;
         while (!SLIST_EMPTY(&modes)) {
             mode = SLIST_FIRST(&modes);
             FREE(mode->name);
 
             /* Clear the old binding list */
-            bindings = mode->bindings;
```
Comment (Airblader) [inline src/config.c]: Just FYI, but I found it rather odd to overwrite a global for the purpose of free'ing memory.

File: src/config.c [hunk 9]
```diff
@@ -75,20 +75,20 @@ void load_configuration(xcb_connection_t *conn, const char *override_configpath,
...
+            while (!TAILQ_EMPTY(mode->bindings)) {
+                Binding *bind = TAILQ_FIRST(mode->bindings);
+                TAILQ_REMOVE(mode->bindings, bind, bindings);
                 binding_free(bind);
             }
-            FREE(bindings);
+            FREE(mode->bindings);
+
             SLIST_REMOVE(&modes, mode, Mode, modes);
+            FREE(mode);
```
Comment (Airblader) [inline src/config.c]: This was a memory leak.

File: src/config.c [hunk 10]
```diff
@@ -110,8 +110,22 @@ void load_configuration(xcb_connection_t *conn, const char *override_configpath,
             FREE(barconfig->id);
             for (int c = 0; c < barconfig->num_outputs; c++)
                 free(barconfig->outputs[c]);
+
+            while (!TAILQ_EMPTY(&(barconfig->bar_bindings))) {
```
Comment (Airblader) [inline src/config.c]: This was a memory leak.
